### PR TITLE
Add support for SMT100 Soil Moisture Sensor

### PR DESCRIPTION
**Pull request by piechade**
# What does this implement/fix?

Add new component for the SMT100 Soil Moisture Sensor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2197

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
    uart:
      tx_pin: TX
      rx_pin: RX
      baud_rate: 9600

    sensor:
      - platform: smt100
        counts:
            name: "Counts"
        permittivity:
            name: "Permittivity"
        temperature:
            name: "Temperature"
        moisture:
            name: "Moisture"
        voltage:
            name: "Voltage"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by piechade** [comment]
> Hey there @piechade, Thanks for submitting this pull request! Can you add yourself as a codeowner for this integration? This way we can notify you if a bug report for this integration is reported. In `__init__.py` of the integration, please add:
> 
> ```python
> CODEOWNERS = ["@piechade"]
> ```
> 
> And run `script/build_codeowners.py`
> 
> (message by NeedsCodeownersLabel)

Done and fixed all other issues from the pipelines.

File: esphome/components/smt100/smt100.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+void SMT100Component::loop() {
+  static char buffer[MAX_LINE_LENGTH];
+  while (this->available() != 0) {
+    if (readline_(read(), buffer, MAX_LINE_LENGTH) > 0) {
+      int counts = (int) strtol((strtok(buffer, ",")), nullptr, 10);
+      float dielectric_constant = (float) strtod((strtok(nullptr, ",")), nullptr);
+      float moisture = (float) strtod((strtok(nullptr, ",")), nullptr);
+      float temperature = (float) strtod((strtok(nullptr, ",")), nullptr);
+      float voltage = (float) strtod((strtok(nullptr, ",")), nullptr);
```
Comment (jesserockz) [inline esphome/components/smt100/smt100.cpp:22]: Not gonna force a change, but this could be simpler using `sscanf`?

**Review by jesserockz — APPROVED** [review]
Just one comment, happy to merge if you want to keep as is though.

File: esphome/components/smt100/sensor.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                device_class=DEVICE_CLASS_VOLTAGE,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+        }
+    )
+    .extend(cv.polling_component_schema("60s"))
+    .extend(uart.UART_DEVICE_SCHEMA)
+)
+
+
```
Comment (jesserockz) [inline esphome/components/smt100/sensor.py:70]: Here you should add a `FINAL_VALIDATE_SCHEMA` that enforces the `uart` pins and baud rate. Search the codebase for examples.

**Comment by piechade** [comment]
> Just one comment, happy to merge if you want to keep as is though.

Thanks for the feedback, It's working, and I'm getting values in home assistant. 
To be honest, I'm not a fan of these one-liners. I have the previous code from the esphome doc.
What would you change on the readline functionß

**Comment by piechade** [comment]
> > Just one comment, happy to merge if you want to keep as is though.
> 
> Thanks for the feedback, It's working, and I'm getting values in home assistant. To be honest, I'm not a fan of these one-liners. I have the previous code from the esphome doc. What would you change on the readline function.

How I get this fixed?
```
Error: /home/runner/work/esphome/esphome/esphome/components/smt100/smt100.cpp:21:7: error: 'sscanf' used to convert a string to an integer value, but function will not report conversion errors; consider using 'strtol' instead [cert-err34-c,-warnings-as-errors]
      sscanf(buffer, "%d,%f,%f,%f,%f", &counts, &dielectric_constant, &moisture, &temperature, &voltage);
```

**Comment by jesserockz** [comment]
Ah I did not realise there was a recommendation against using `sscanf`. Your original code there is fine then.

**Comment by piechade** [comment]
> Ah I did not realise there was a recommendation against using `sscanf`. Your original code there is fine then.

Should I use vector instead?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**